keps/sig-instrumentation/1753-logs-sanitization/README.md

```diff
@@ -0,0 +1,243 @@
+
+This risk can be mitigated by:
+- running inspection only on log entry parameters which are going to be actually logged - running inspection after log v-level has been evaluated.
+- introducing a dedicated go lang interface with a single method returning information about types of sensitive data which given value contains like:
+
+  ```go
+  type DatapolicyDetector interface {
+	  ContainsDatapolicyValues() []string
+  }
+  ```
+
+- implementations of this interface could be auto generated using a dedicated code generator similar to deep-copy generator or manually implemented when needed,
+caching negative inspection results for parameter types which does not have any references to types which may contain sensitive data.
+
+Which of those methods will be used and to what extent will be decided after running performance tests.
+
+
+## Design Details
+
+### Source code tags
+
+We propose to mark all struct fields which may contain sensitive data with a new datapolicy tag which as a value will accept a list of predefined types of data.
+
+For now we propose following types of data to be available as values:
+- password
+- token
+- security-key
+
+Example of using datapolicy tag:
+
+```go
+type ConfigMap struct {
+  Data map[string]string `json:"data,omitempty" datapolicy:”password,token,security-key”`
+}
+```
+
+For external types which are not part of the Kubernetes source code such as go lang standard libraries or third party vendored libraries for which we cannot change source code we will define a global function which will map those into relevant datapolicy tag:
+
+```go
+func GlobalDatapolicyMapping(val interface{}) []string
+```
+
+### datapolicy verification library
+
+datapolicy verification library will implement verification if the provided value does not contain any sensitive data identified by the datapolicy tag. Verification will be performed using reflection. Recursion will stop on pointers as values for those are usually not logged. Validation will depend only on the datapol tags values. Values of individual fields of primitive types like string will not be analyzed other than checking if a given field is not empty.
+
+```go
+package datapol {
+  // Verify returns a slice with types of sensitive data the given value contains. 
+  func Verify(value interface{}) []string
+}
+```
+
+When used datapolicy verification library will be initialized with GlobalDatapolicyMapping function to take into account information about external types which may contain sensitive data.
+
+### klog integration
+
+Currently the klog library used by Kubernetes does not provide any extension point which could be used for filtering logs.
+
+We propose to add a new interface to klog library:
+
+```go
+type LogFilter interface {
+  Filter(args []interface{}) (args []interface{})
+  FilterF(format string, args []interface{}) (string,[]interface{})
+  FilterS(msg string, keysAndValues []interface{}) (string, []interface{})
+}
```

Comment:
Why are we exposing this interface?

Reply:
We do not want to implement sanitization logic as an integral part of the klog but instead we want to inject it whenever the Kubernetes component is configured to do so. Therefore we need to expose this interface so that it could be implemented in Kubernetes.